connection.go

```diff
@@ -1233,6 +1370,212 @@ func (conn *Connection) NewStream() (*Stream, error) {
 	}, nil
 }
 
+// watchState is the current state of the watcher. See the idea at p. 70, 105:
+// https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view
+type watchState struct {
+	// value is a current value.
+	value interface{}
+	// version is a current version of the value. The only reason for uint64:
+	// go 1.13 has no math.Uint.
+	version uint64
+	// ack true if the acknowledge is already sended.
+	ack bool
+	// cnt is a count of active watchers for the key.
+	cnt int
+	// changed is a channel for broadcast the value changes.
+	changed chan struct{}
+	// unready channel exists if a state is not ready to work (subscribtion
+	// or unsubscribtion in progress).
+	unready chan struct{}
+}
+
+// initWatchEventVersion is an initial version until no events from Tarantool.
+const initWatchEventVersion = 0
+
+// connWatcher is an internal implementation of the Watcher interface.
+type connWatcher struct {
+	unregister sync.Once
+	done       chan struct{}
+	finished   chan struct{}
+}
+
+// Unregister unregisters the connection watcher.
+func (w *connWatcher) Unregister() {
+	w.unregister.Do(func() {
+		close(w.done)
+	})
+	<-w.finished
+}
+
+// subscribeWatchChannel returns an existing one or a new watch state channel
+// for the key. It also increases a counter of active watchers for the channel.
+func subscribeWatchChannel(conn *Connection, key string) (chan watchState, error) {
+	var st chan watchState
+
+	for st == nil {
+		if val, ok := conn.watchMap.Load(key); !ok {
+			st = make(chan watchState, 1)
+			state := watchState{
+				value:   nil,
+				version: initWatchEventVersion,
+				ack:     false,
+				cnt:     0,
+				changed: nil,
+				unready: make(chan struct{}),
+			}
+			st <- state
+
+			if val, ok := conn.watchMap.LoadOrStore(key, st); !ok {
+				if _, err := conn.Do(newWatchRequest(key)).Get(); err != nil {
+					conn.watchMap.Delete(key)
+					close(state.unready)
+					return nil, err
+				}
+				// It is a successful subsctibtion to a watch events by itself.
+				state = <-st
+				state.cnt = 1
+				close(state.unready)
+				state.unready = nil
+				st <- state
+				continue
+			} else {
+				close(state.unready)
+				close(st)
+				st = val.(chan watchState)
+			}
+		} else {
+			st = val.(chan watchState)
+		}
+
+		// It is an existing channel created outside. It may be in the
+		// unready state.
+		state := <-st
+		if state.unready == nil {
+			state.cnt += 1
+		}
+		st <- state
+
+		if state.unready != nil {
```

Comment:
Isn't it possiblt to use a single if/else? For now it seems that there is a minor duplication of logic.

Reply:
```Go
if state.unready != nil {
    state.cnt += 1
    st <- state
    <-state.unredy
    st = nil
} else {
    st <- state
}
```

It doesn't look much simpler, but breaks the workflow with the channel:

```Go
state := <- st
// some light change or check
st <- state
```

Reply:
Why it's not a least a

```
if state.unready == nil {
    state.cnt += 1
    st <- state
}
```

Reply:
My fault, so we have two alternatives:
```Go
state := <-st
if state.unready == nil {
	state.cnt += 1
}
st <- state

if state.unready != nil {
	// Wait for an update and retry.
	<-state.unready
	st = nil
}
```
vs
```Go
state := <-st
if state.unready == nil {
    state.cnt += 1
    st <- state
} else {
    st <- state
    <-state.unredy
    st = nil
}
```

I think the first one is easier to parse with your eyes and find the pattern.

Reply:
Oh, we need to put state back even if it wasn't changed. I think I get it now